File: chronos-backend/src/main/java/net/fvogel/chronosbackend/admin/bulkactions/addtags/AddTagsBulkActionService.java
package net.fvogel.chronosbackend.admin.bulkactions.addtags;

import net.fvogel.chronosbackend.common.persistence.entries.model.Entry;
import net.fvogel.chronosbackend.common.persistence.tags.model.Tag;
import net.fvogel.chronosbackend.common.service.EntriesService;
import net.fvogel.chronosbackend.common.service.TagsService;
import org.springframework.stereotype.Service;

import java.util.Set;

@Service
public class AddTagsBulkActionService {

    private TagsService tagsService;
    private EntriesService entriesService;

    AddTagsBulkActionService(
            TagsService tagsService,
            EntriesService entriesService) {
        this.tagsService = tagsService;
        this.entriesService = entriesService;
    }

    public void addTagsToEntries(Set<Long> tagIds, Set<Long> entryIds) {
        Set<Tag> tags = this.tagsService.findByIds(tagIds);
        Set<Entry> entries = this.entriesService.findByIds(entryIds);
        for (Entry entry : entries) {
            entry.getTags().addAll(tags);
        }
        this.entriesService.saveAll(entries);
    }
}

File: chronos-frontend/src/app/modules/admin/services/bulk-actions/add-tags-bulk-action.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from "@angular/common/http";
import {Observable} from "rxjs";
import {Tag} from "../../../../model/tag.model";
import {Entry} from "../../../../model/entry.model";

interface AddTagsBulkAction {
  tagIds: Array<number>,
  entryIds: Array<number>
}


@Injectable({
  providedIn: null
})
export class AddTagsBulkActionService {

  constructor(private http: HttpClient) {
  }

  public addTagsToEntries(tags: Array<Tag>, entries: Array<Entry>): Observable<void> {
    const body: AddTagsBulkAction = {
      tagIds: tags.filter(t => !!t.id).map(t => t.id!),
      entryIds: entries.filter(e => !!e.id).map(e => e.id!)
    };
    return this.http.put<void>('/api/admin/bulk-actions/entries/add-tags', body)
  }

}
